[RFC] Generic ARM64 under Xen with serial and block frontend

andr2000 opened the pull request:
This is for pre-upstream review and comments

Use `make xenguest_arm64_defconfig` for configuration

Commit:
board: Introduce xenguest_arm64 board

Introduce a minimal Xen guest board running as a virtual
machine under Xen Project's hypervisor [1], [2].

Known issues:
 - memory size cannot be less roughly 129M as Xen puts
   the device tree at 0x48000000 and RAM starts 0x40000000
 - getc is not implemented as it requires para-virtualized
   console driver to be implemented
 - no prompt is displayed as puts we use expects "\n" at the
   end to print the string which is not the case with the prompt.

Part of the code is ported from Xen mini-os and also uses
work initially done by different authors from NXP: please see
relevant files for their copyrights.

[1] https://xenbits.xen.org
[2] https://wiki.xenproject.org/

Signed-off-by: Andrii Anisov <andrii_anisov@epam.com>
Signed-off-by: Oleksandr Andrushchenko <oleksandr_andrushchenko@epam.com>

Commit:
lib: sscanf: add sscanf implementation

Port sscanf implementation from mini-os and introduce new
Kconfig option to enable it: CONFIG_SSCANF. Disable by default.

Signed-off-by: Andrii Anisov <andrii_anisov@epam.com>

Commit:
xen: Port Xen bus driver from mini-os

Make required updates to run on u-boot and strip test code.

Commit:
xen: Port Xen grant table driver from mini-os

Make required updates to run on u-boot.

Signed-off-by: Oleksandr Andrushchenko <oleksandr_andrushchenko@epam.com>

Commit:
linux/compat.h: Add wait_event_timeout macro

Add  wait_event_timeout - sleep until a condition gets true or a
timeout elapses.

This is a stripped version of the same from Linux kernel with the
following u-boot specific modifications:
- no wait queues supported
- use u-boot timer to detect timeouts
- check for Ctrl-C pressed during wait

Signed-off-by: Oleksandr Andrushchenko <oleksandr_andrushchenko@epam.com>

MrVan committed:
Kconfig: Introduce CONFIG_XEN

Introduce CONFIG_XEN to make U-Boot could be used as bootloader
for a virtual machine.

Without bootloader, we could successfully boot up android on XEN, but
we need need bootloader to support A/B, dm verify and etc.

Signed-off-by: Peng Fan <peng.fan@nxp.com>
Signed-off-by: Oleksandr Andrushchenko <oleksandr_andrushchenko@epam.com>

Commit:
xen: include XEN interface headers from RELEASE-4.13.1

Add Xen interface headers from Xen RELEASE-4.13.1, at
commit 6278553325a9f76d37811923221b21db3882e017

Signed-off-by: Oleksandr Andrushchenko <oleksandr_andrushchenko@epam.com>

Commit:
[HACK][DO NOT MERGE]: Move RAM base to 80000000

Signed-off-by: Oleksandr Andrushchenko <oleksandr_andrushchenko@epam.com>

Commit:
Fix SMCC and ARM_PSCI_FW dependencies

Currently SMCC selects ARM_PSCI_FW if enabled which is not correct
as there are cases that PSCI can function without firmware at all.
ARM_PSCI_FW itself is built with driver model approach, so it cannot
be enabled if DM is off.
Fix this by making PSCI reset functionality depend on ARM_PSCI_FW and
only in case if DM is enabled.

Signed-off-by: Oleksandr Andrushchenko <oleksandr_andrushchenko@epam.com>
Suggested-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

Commit:
xen: Port Xen hypervizor related code from mini-os

Port hypervizor related code from mini-os. Update essential
arch code to support required bit operations, memory barriers etc.

Copyright for the bits ported belong to at least the following authors,
please see related files for details:

Copyright (c) 2002-2003, K A Fraser
Copyright (c) 2005, Grzegorz Milos, gm281@cam.ac.uk,Intel Research Cambridge

Signed-off-by: Oleksandr Andrushchenko <oleksandr_andrushchenko@epam.com>

MrVan committed:
serial: serial_xen: Add Xen PV serial driver

Add support for Xen para-virtualized serial driver. This
driver fully supports serial console for the virtual machine.

Please note that as the driver is initialized late, so no banner
nor memory size is visible.

Signed-off-by: Peng Fan <peng.fan@nxp.com>
Signed-off-by: Oleksandr Andrushchenko <oleksandr_andrushchenko@epam.com>

Commit:
xen: Port Xen event channel driver from mini-os

Make required updates to run on u-boot.

Signed-off-by: Oleksandr Andrushchenko <oleksandr_andrushchenko@epam.com>

Commit:
xen: pvblock: Add initial support for para-virtualized block driver

Add initial infrastructure for Xen para-virtualized block device.
This includes compile-time configuration and the skeleton for
the future driver implementation.
Add new class UCLASS_PVBLOCK which is going to be a parent for
virtual block devices.
Add new interface type IF_TYPE_PVBLOCK.

Implement basic driver setup by reading XenStore configuration.

Signed-off-by: Andrii Anisov <andrii_anisov@epam.com>
Signed-off-by: Anastasiia Lukianenko <anastasiia_lukianenko@epam.com>
Signed-off-by: Oleksandr Andrushchenko <oleksandr_andrushchenko@epam.com>

Commit:
configs: Add ext4 and fat commands for xenguest_arm64

Signed-off-by: Anastasiia Lukianenko <anastasiia_lukianenko@epam.com>

Commit:
xen: pvblock: Enumerate virtual block devices

Enumerate Xen virtual block devices found in XenStore and
instantiate pvblock devices.

Signed-off-by: Oleksandr Andrushchenko <oleksandr_andrushchenko@epam.com>
Signed-off-by: Anastasiia Lukianenko <anastasiia_lukianenko@epam.com>

Commit:
xen: pvblock: Read XenStore configuration and initialize

Read essential virtual block device configuration data from XenStore,
initialize front ring and event channel.
Update block device description with actual block size.

Use code for XenStore from mini-os.

Signed-off-by: Oleksandr Andrushchenko <oleksandr_andrushchenko@epam.com>
Signed-off-by: Anastasiia Lukianenko <anastasiia_lukianenko@epam.com>

Commit:
xen: pvblock: Implement front-back protocol and do IO

Implement Xen para-virtual frontend to backend communication
and actually read/write disk data.

This is based on mini-os implementation of the para-virtual block
frontend driver.

Signed-off-by: Oleksandr Andrushchenko <oleksandr_andrushchenko@epam.com>
Signed-off-by: Anastasiia Lukianenko <anastasiia_lukianenko@epam.com>

lorc committed:
arm64: issue ISB after updating system registers

ARM Architecture reference manual clearly states that PE pipeline
should be flushed after any change to system registers. Failing to
do so can lead to spurious errors, like synchronous exception when
accessing FPU registers. This is very prominent on CPUs with long
instruction pipeline, like Cortex A72.

This change fixes the following U-Boot panic:

 "Synchronous Abort" handler, esr 0x1fe00000
 elr: 00000000800948cc lr : 0000000080091e04
 x0 : 00000000801ffdc8 x1 : 00000000000000c8
 x2 : 00000000800979d4 x3 : 00000000801ffc60
 x4 : 00000000801ffd40 x5 : ffffff80ffffffd8
 x6 : 00000000801ffd70 x7 : 00000000801ffd70
 x8 : 000000000000000a x9 : 0000000000000000
 x10: 0000000000000044 x11: 0000000000000000
 x12: 0000000000000000 x13: 0000000000000000
 x14: 0000000000000000 x15: 0000000000000000
 x16: 000000008008b2e0 x17: 0000000000000000
 x18: 00000000801ffec0 x19: 00000000800957b0
 x20: 00000000000000c8 x21: 00000000801ffdc8
 x22: 000000008009909e x23: 0000000000000000
 x24: 0000000000000000 x25: 0000000000000000
 x26: 0000000000000000 x27: 0000000000000000
 x28: 0000000000000000 x29: 00000000801ffc50

 Code: a94417e4 a90217e4 a9051fe6 a90617e4 (3d801fe0)

While executing instruction

 str     q0, [sp, #112]

in vsnprintf() prologue. This panic was observed on Cortex A72 so far.

Also, this probably the right fix for the issue workarounded in the
commit 45f41c13 ("ARM: uniphier: add weird workaround code for LD20")

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

Commit:
xen: pvblock: Print found devices indices

Signed-off-by: Oleksandr Andrushchenko <oleksandr_andrushchenko@epam.com>
Signed-off-by: Anastasiia Lukianenko <anastasiia_lukianenko@epam.com>

jgrall commented on arch/arm/Makefile:
```diff
@@ -28,6 +28,10 @@ else
 endif
 
+ifeq ($(CONFIG_XEN),y)
+arch-y += -D__XEN__
```
__XEN__ is meant to be used by the hypervisor only. It protects anything that is not meant to be stable from a domain PoV and will default the ABI to the latest one. This will have an impact longer term as you update the headers. As U-boot is running in a domain, he needs to use the stable ABI.


As U-boot will run in a domain, it should only used the stable ABI. AFAICT, you are

andr2000 replied:
Yes, I understand that both __XEN__ and __XEN_TOOLS__ are there for good purpose. But to me it had cons and pros of either hardcoding the values or using __XEN__ to get those from the relevant header

jgrall replied:
I think you missed my point. By using anything guarded by those define, you effectively tie U-boot to a specific version of Xen. To give a concrete example, if tomorrow we decide to rework the memory layout), then U-boot is going to break.

How do you plan to deal with that in upstream U-boot?

andr2000 replied:
I understand that and it seems I didn't miss the point. Yes, we do tie to this specific version of Xen by copying its headers and using __XEN__. But what else can we do? Define those constants WRT, again, to the version of Xen we are running on. So, both ways will require changing either Xen header(s) to be up to date with Xen or changing the relevant constant in `include/configs/xenguest_arm64.h`

jgrall replied:
Well, when you want to do the right thing from scratch, it will always take some effort...
I can see another alternative. I am not happy with it but I will mention it for fairness.
You could require the user to provide which version of Xen he/she is using. This would be used to define the list of RAM banks used.

Commit:
serial: serial_xen: Enable prints at early stages

We may be running with d-cache off at some initial boot stages,
but according to Xen hypercall ABI (see include/public/arch-arm.h)
all the shared buffers must reside in memory which is mapped as
Normal Inner Write-Back Inner-Shareable. This means that before the
MMU is set up we need to perform dcache maintenance manually.

To see serial console prints before the MMU set up one needs to
enable debug and verbose debug messages in Xen with CONFIG_VERBOSE_DEBUG
configuration option: those prints include, in particular, u-boot
banner, CPU and machine model and DRAM size.

Signed-off-by: Oleksandr Andrushchenko <oleksandr_andrushchenko@epam.com>
Signed-off-by: Anastasiia Lukianenko <anastasiia_lukianenko@epam.com>

jgrall commented on include/configs/xenguest_arm64.h:
```diff
@@ -0,0 +1,47 @@
+ * See for reference:
+ * https://xenbits.xen.org/gitweb/?p=xen.git;a=blob;f=xen/include/public/arch-arm.h
+ */
+#define CONFIG_NR_DRAM_BANKS          GUEST_RAM_BANKS
```
Neither **GUEST_RAM_BANKS** and **GUEST_RAM_BASE0** are part of the stable ABI. So you can't safely rely on those values in U-boot.

I am not very familiar with U-boot, would you mind to explain why does U-boot requires to know the physical address? Why can't it find out from the device-tree?

andr2000 replied:
This is because of the linker script mostly which needs to know where to put the sections

jgrall replied:
This makes sense for ``GUEST_RAM_BASE0`` but why would the linker need ``GUEST_RAM_BANKS``?

However, I am wondering whether you can make the code PIE?

andr2000 replied:
GUEST_RAM_BANKS is needed for the u-boot itself to iterate over the predefined number of banks. Basically, you need to set CONFIG_NR_DRAM_BANKS for u-boot to be happy.
You can grep the code for the usages of that to see more
WRT PIE I am not quite sure this is an easy task for u-boot at the moment. Neither this is the target of the pull request

jgrall replied:
See my comment https://github.com/xen-troops/u-boot/pull/1#discussion_r442093400.

You are tying U-boot to a specific version of Xen. This is potentially even a per-guest configuration depending on whether we decide to expose the host layout to the guest.

Do we really want to go down the road where you have to a new U-boot binary for each of your guest?

andr2000 replied:
Yes, but taking into account wjay u-boot expects from us I can't tell how to deal with that in some generic way with the existing code base without major changes

jgrall replied:
The short answer is don't use anything under __XEN__. If you are planning to upstream this in U-boot, you will not make a favor to any of the user.

It would definitely be best to define a list of supported hypervisor in U-boot or try to work with Xen community to guarantee the position of the first bank in memory.

andr2000 replied:
I see another option as well:
1. Put the headers outside the u-boot tree and provide their location.
This way we can build against different versions of Xen without changing u-boot code
2. We can have a set of the relevant constants in xenguest_arm64.h surrounded with 
`#if __XEN_INTERFACE_VERSION__ < XXX
` with the version provided from outside in some way

andr2000 replied:
Or we can setup the relevant addresses vi Kconfig, but number of banks depends on live dtb, so need a predefined value anyways

jgrall replied:
> I see another option as well:
> 
> 1. Put the headers outside the u-boot tree and provide their location.
>    This way we can build against different versions of Xen without changing u-boot code

This relies on the naming scheme to never change and that we will never provide more flexibility to the user on the layout. So it is not very ideal.

> 2. We can have a set of the relevant constants in xenguest_arm64.h surrounded with
>    `#if __XEN_INTERFACE_VERSION__ < XXX ` with the version provided from outside in some way

This would look sensible. Although, this relies on end user to hack arch-arm.h as you did in your branch. So you would want to allow the user to provide there own easily.

I will take the opportunity to note that Tianocore/EDK2 on Xen is pretty simple to use as you can compile once and then run it on every existing Xen. At the moment, all the options suggested make U-boot much less accessible to the users. This may be fine if you hack it in your side, but for a general use you need to find an easy way for the users to get things reliably working.

jgrall replied:
> Or we can setup the relevant addresses vi Kconfig, but number of banks depends on live dtb, so need a predefined value anyways

This is a recipe for disaster. Why does U-boot need to know all the banks?

sstabellini replied:
Uboot is one of those embedded projects that typically assumes that all the information about the platform is available at *build time*. It is meant to be built tailored for a specific version of a specific board. A Uboot binary is not expected to be "portable" across different versions of the platform or different platforms. In other words, it is not expected to be portable across Xen versions.

So I don't think it is a problem to let Uboot know about Xen details at build time. Yes, that is not what we do historically but it is very much in the spirit of Uboot.

I think the question we should ask ourselves is what is the best and most flexible way to expose this information to UBoot. Xen headers? Kconfig? Device Tree (Uboot sometimes take a device tree at build time for configuration, not run time).

I don't know for sure but I suspect that using the Xen headers is better than using a Kconfig.

andr2000 replied:
Minimal set of headers + only stable ABI definitions

jgrall commented on board/xen/xenguest_arm64/xenguest_arm64.c:
```diff
@@ -0,0 +1,158 @@
+	 * Add "magic" region which is used by Xen to provide some essentials
+	 * for the guest: console, xenstore etc.
+	 */
+	xen_mem_map[0].virt = GUEST_MAGIC_BASE;
```
I was going to ask how you deal with the cache here, but it looks like _dcache_enable()_ will flush the cache entirely with set/way :(.

I would strongly suggest to avoid set/way as this is a really expensive operation in virtualization (we have to walk the guest memory and flush line by line).

andr2000 replied:
This is probably what I see that if given more memory the guests starts slowly

jgrall commented on drivers/serial/serial_xen.c:
```diff
@@ -0,0 +1,79 @@
+	.stop	= NULL,
+	.setbrg	= xen_serial_setbrg,
+	/*
+	 * FIXME: Non-priveleged domain should not try reading characters via
```
An unprivileged guest cannot use _CONSOLEIO_write()_ on production build (unless you implement your own XSM flask). If you expect U-boot to print important message using this interface, then I would recommend to use a different interface (either the PV console or the pl011).

andr2000 replied:
Well, this piece of code is only used if driver model is not enabled and intention was to provide this interface **before** we implement PV console. So, at the end of the day PV console is in place

jgrall replied:
I am not sure to understand this... Is the code in use or not? If it is in use what kind of message do you print on it?

andr2000 replied:
It is in use if driver model is disabled which is the case for the very first patch. After we add more meat we enable the driver model and PV console is in charge then.

andr2000 replied:
Basically we can drop that code at the end

jgrall replied:
A general of rule is we don't add feature that will be dropped by the end of the series. Although, I don't know what's U-boot stance for such things.

andr2000 replied:
We have dropped the serial driver without the driver model

Commit:
common/board_f: Respect original FDT size while relocating

While relocating FDT we reserve some memory for the new FDT and
set the size of the FDT with that respect. But FDT may be placed
at the end of the RAM leading to memory access beyond it.
Fix this by copying exact FDT size bytes, not the reserved size.

Signed-off-by: Oleksandr Andrushchenko <oleksandr_andrushchenko@epam.com>

jgrall commented on drivers/xen/gnttab.c:
```diff
@@ -0,0 +1,220 @@
+    for (i = NR_RESERVED_ENTRIES; i < NR_GRANT_ENTRIES; i++)
+        put_free_entry(i);
+
+    gnttab_table = memalign(PAGE_SIZE, NR_GRANT_FRAMES * PAGE_SIZE);
```
AFAIU, this will allocate part of the RAM for the grant-table. Will this region be marked as unusable for the DomU OS afterwards?

andr2000 replied:
When jumping to the kernel we remove all the mappings, so the kernel will setup its own gnttab, so no need to mark any region

jgrall replied:
Would you mind to give me a pointer to that code removing the grant table frames?

andr2000 replied:
https://github.com/xen-troops/u-boot/pull/1/commits/f065d0371da9e15e19fd91b9d71604e00814ff6c#diff-38af42740ed5d67089e2992db6a1a39cR209
Please note that `fini_gnttab` is not called in this pull request, but it is in the new one

jgrall replied:
Well, the hypercall **GNTTABOP_setup_table** doesn't behave as you think. For Arm, it will only ensure there is enough frames allocated in the hypervisor.  There is no management of the mappings in the p2m.

This is the guest job to map/unamp them using the appropriate hypercalls. The grant-table frames as the vCPU shared pages are memory from Xen mapped into the guest address space. So the proper way for map/unmap is the following:

-  For the mapping 
1. Remove the RAM mapping using **XENMEM_decrease_reservation**
2. Map the grant frame using **XENMEM_add_to_physmap**

- For the unmapping
1. Unmap the grant frame using **XENMEM_remove_from_physmap**
2. Populate the region using **XENMEM_populate_physmap**

Step 1 of the mapping is required to avoid losing the RAM page for the lifetime of your guest.
Step 2 of the unmapping is required so the address point again to a RAM page.

andr2000 replied:
Probably I miss your point, but we do not map and unmap any foreign grant references, we only share our own pages with the backend.
For the shared page we use:
```
	xatp.space = XENMAPSPACE_shared_info;
	xatp.gpfn = virt_to_pfn(HYPERVISOR_shared_info);
	if (HYPERVISOR_memory_op(XENMEM_add_to_physmap, &xatp) != 0)

```
For the grant-table we use:
```
        xatp.space = XENMAPSPACE_grant_table;
        xatp.gpfn = ((unsigned long)gnttab_table >> PAGE_SHIFT) + i;
        rc = HYPERVISOR_memory_op(XENMEM_add_to_physmap, &xatp);
...
    setup.nr_frames = NR_GRANT_FRAMES;
    set_xen_guest_handle(setup.frame_list, frames);
    rc = HYPERVISOR_grant_table_op(GNTTABOP_setup_table, &setup, 1);

```
When we setup a shared ring we:
```
	SHARED_RING_INIT(s);
...
	dev->ring_ref = gnttab_grant_access(dev->dom, virt_to_pfn(s), 0);

```
While pvblock shares the pages we use:
```
			gnttab_grant_access(...),
			gnttab_end_access(...);

```
So, as we do not have any foreign mappings, why would we need anything else like you mentioned?

jgrall replied:
> Probably I miss your point, but we do not map and unmap any foreign grant references, we only share our own pages with the backend.

Notice that I mentioned memory shared with **xen** and not the **backend**. The way the two works are fairly different see more below.

>
> For the shared page we use:
> 
> ```
> 	xatp.space = XENMAPSPACE_shared_info;
> 	xatp.gpfn = virt_to_pfn(HYPERVISOR_shared_info);
> 	if (HYPERVISOR_memory_op(XENMEM_add_to_physmap, &xatp) != 0)
> ```

For this one and ...

> 
> For the grant-table we use:
> 
> ```
>         xatp.space = XENMAPSPACE_grant_table;
>         xatp.gpfn = ((unsigned long)gnttab_table >> PAGE_SHIFT) + i;
>         rc = HYPERVISOR_memory_op(XENMEM_add_to_physmap, &xatp);
> ```

... this one, they are used to map **xen** memory into the guest. IOW, the P2M entry associated to the field *xatp.gfn* will be updated to point to **xen** memory. If there was any mapping (such as to a RAM page), then it will disappear forever. Effectively, you will leave a hole in your address space.

But I would strongly recommend to avoid using RAM page for the grant-table if you can. This will end up to shatter superpages in your P2M and potentially affect the performance of your guest long-term. You will be better off to use the region provided by the device-tree.

andr2000 replied:
Ok, this can be done (grant table from the device tree). And if we use those then we don't need to play with reservations, right?

jgrall replied:
You will not need to remove the previous mapping and repopulate it, but you will still have to unmap whatever you map to stay on the _safe_ side.

andr2000 replied:
Thank you, we have implemented the gnttab with the use of device tree reg and XENMEM_add_to_physmap/XENMEM_remove_from_physmap and without XENMEM_decrease_reservation/XENMEM_populate_physmap.
Much cleaner now.

iusyk commented on arch/arm/include/asm/xen/hypercall.h:
```diff
@@ -0,0 +1,68 @@
+int HYPERVISOR_platform_op_raw(void *arg);
+int HYPERVISOR_xsm_op(struct xen_flask_op *);
+
+static inline int HYPERVISOR_platform_op(struct xen_platform_op *op)
```
What is the sense of 'static inline' in.h file ?
Really, it make a sense if function is in .c/cpp file.

jgrall replied:
> What is the sense of 'static inline' in.h file ?

The behavior between C and C++ is quite different when defining inline helper in a header.

In the case of C, if you don't add **static**, the compiler will consider this is a prototype and ignore the body and it will result to a linking error. Below an example:

```
42sh> cat test.c
#include "test.h"

int main(void)
{
	return static_inline() + inline_only();
}
```
```
42sh> cat test.h
#ifndef __TEST_H__

inline int inline_only(void)
{
	return 1;
}

static inline int static_inline(void)
{
	return 2;
}

#endif /* !__TEST_H__ */
```

```
42sh> gcc test.c
/usr/bin/ld: /tmp/ccyxE8AD.o: in function `main':
test.c:(.text+0x1c): undefined reference to `inline_only'
collect2: error: ld returned 1 exit status
```

For C++, the compiler will just create a weak reference.

> Really, it make a sense if function is in .c/cpp file.

I would actually argue the inverse. _static inline_ only makes sense in header because the compiler will often do a better job than you to decide what to inline (_static_ function can be inlined). So unless you specify, **__attribute__((always_inline))**, your compiler is free to ignore your request to inline it.

Note that _static inline_ is still necessary in the header because without _inline_ the compiler will throw a warning if the helper is not used and compile with _-Wall_/_-Wunused-function_

iusyk replied:
ok, thanks

iusyk commented on arch/arm/include/asm/xen/system.h:
```diff
@@ -0,0 +1,86 @@
+})
+
+/* As test_and_clear_bit, but using __ATOMIC_SEQ_CST */
+static inline int synch_test_and_clear_bit(int nr, volatile void *addr)
```
the same like above - no sense to have static here

iusyk commented on drivers/xen/events.c:
```diff
@@ -0,0 +1,258 @@
+	struct vcpu_info *vcpu_info = &s->vcpu_info[cpu];
+
+	for (i = 0; i < NR_EVS; i++) {
+		if (i == console_evtchn)
```
Maybe it is reasonable to reduce the amount of code 
		if (i != console_evtchn && test_and_clear_bit(i, bound_ports)) {
			printf("port %d still bound!\n", i);
			unbind_evtchn(i);
		}

iusyk commented on drivers/xen/pvblock.c:
```diff
@@ -0,0 +1,813 @@
+	}
+
+	RING_FINAL_CHECK_FOR_RESPONSES(&dev->ring, more);
+	if (more)
```
I propose use do{}while(more) instead of goto

iusyk commented on drivers/xen/pvblock.c:
```diff
@@ -0,0 +1,813 @@
+		sscanf(dirs[i], "%d", &devid);
+		ret = clb(udev, devid);
+		if (ret < 0)
+			goto fail;
```
i think break is enough here (to avoid goto)

iusyk commented on drivers/xen/xenbus.c:
```diff
@@ -0,0 +1,532 @@
+			 MASK_XENSTORE_IDX(xenstore_buf->rsp_cons),
+			 sizeof(msg));
+
+	if (xenstore_buf->rsp_prod - xenstore_buf->rsp_cons < sizeof(msg) + msg.len)
```
use do{}while(xenstore_buf->rsp_prod - xenstore_buf->rsp_cons < sizeof(msg) + msg.len)

iusyk commented on drivers/xen/xenbus.c:
```diff
@@ -0,0 +1,532 @@
+			xbt_flag = 1;
+		}
+
+		msg = xenbus_read(xbt, path, &current_state);
```
it is possible to avoid the goto , one of the option ,
if (!msg)
{
	rs = (XenbusState) (current_state[0] - '0');
	free(current_state);
	if (rs != state) {
		snprintf(value, 2, "%d", state);
		msg = xenbus_write(xbt, path, value);
        }
        else
        {
		msg = NULL;
	}

}

iusyk commented on include/xen/events.h:
```diff
@@ -0,0 +1,47 @@
+int evtchn_alloc_unbound(domid_t pal, evtchn_handler_t handler,
+			 void *data, evtchn_port_t *port);
+
+static inline int notify_remote_via_evtchn(evtchn_port_t port)
```
static is not required here, what is the idea of static ?
